cpp/include/rapidsmpf/buffer/packed_data.hpp

```diff
@@ -22,39 +24,39 @@ namespace rapidsmpf {
  */
 struct PackedData {
     std::unique_ptr<std::vector<std::uint8_t>> metadata;  ///< The metadata
-    std::unique_ptr<rmm::device_buffer> gpu_data;  ///< The gpu data
+    std::unique_ptr<Buffer> gpu_data;  ///< The gpu data
+    /// @brief Default constructor
+    PackedData() = default;
 
     /**
-     * @brief Construct packed data from metadata and gpu data, taking ownership.
+     * @brief Construct from metadata and gpu data.
      *
-     * @param meta The metadata
-     * @param data The gpu data
+     * @param metadata Host-side metadata describing the GPU data.
+     * @param gpu_data Pointer to GPU data.
      */
     PackedData(
-        std::unique_ptr<std::vector<std::uint8_t>>&& meta,
-        std::unique_ptr<rmm::device_buffer>&& data
+        std::unique_ptr<std::vector<std::uint8_t>> metadata,
+        std::unique_ptr<Buffer> gpu_data
     )
-        : metadata{std::move(meta)}, gpu_data{std::move(data)} {
-        RAPIDSMPF_EXPECTS(
-            metadata != nullptr || gpu_data != nullptr,
-            "Metadata or GPU data must be non-null"
-        );
-
-        RAPIDSMPF_EXPECTS(
-            (metadata->size() > 0 || gpu_data->size() == 0),
-            "Empty Metadata and non-empty GPU data is not allowed"
-        );
-    }
+        : metadata{std::move(metadata)}, gpu_data{std::move(gpu_data)} {}
 
     /**
-     * @brief Construct an empty PackedData object.
+     * @brief Construct from metadata and device buffer, taking ownership.
      *
-     * This constructor initializes both the metadata and GPU data to empty
-     * buffers.
+     * @param metadata Host-side metadata.
+     * @param gpu_data Raw GPU buffer to be moved into a wrapped buffer.
+     * @param stream CUDA stream used during the move.
+     * @param br Buffer resource for gpu data allocations.
+     * @param event Optional event to associate with the gpu data.
      */
-    PackedData()
-        : metadata{std::make_unique<std::vector<std::uint8_t>>()},
-          gpu_data{std::make_unique<rmm::device_buffer>()} {}
+    PackedData(
+        std::unique_ptr<std::vector<std::uint8_t>> metadata,
+        std::unique_ptr<rmm::device_buffer> gpu_data,
+        rmm::cuda_stream_view stream,
+        BufferResource* br,
+        std::shared_ptr<Buffer::Event> event = nullptr
+    )
+        : PackedData(std::move(metadata), br->move(std::move(gpu_data), stream, event)) {}
```

Comment:
Since the device buffer is on GPU, is the idea with this move that we _might_ spill to host (hence the stream/event)?